### ISSUE TITLE
(aws-networkfirewall): unable to set IpSets in RuleVariables

### ISSUE THREAD

**Issue by jameszh** [aws/aws-cdk#15375]
not able to set IpSets in RuleVariables

### Reproduction Steps

The following is the snippet in python 

```python
           rule_group={
                "ruleVariables": {
                    "iPSets": {
                        "HOME_SET": {
                            "definition": [
                                "192.168.0.0/24"
                            ]
                        }
                    },
                    "portSets": {
                        "MYSQL": {
                            "definition": [
                                "3306"
                            ]
                        }
                    }
                },

```

### What did you expect to happen?

It's deployed successfully. When checking with describe-rule-group using cli,  IpSets are not shown, only PortSets are in the output.

I  expect IpSets are created as below

```json
    "RuleVariables": {
        "IPSets": {
            "HOME_SET": {
                "Definition": [
                    "10.160.0.0/24"
                ]
            }
        }
    },

```
### What actually happened?
IPSets not created

### Environment

  - **CDK CLI Version  :**1.110.1 (build 0028d09)
  - **Framework Version:**
  - **Node.js Version:**v14.15.5
  - **OS               :**macOS 11.4
  - **Language (Version):**Python 3.9.5

### Other

N/A

--- 

This is :bug: Bug Report

**Comment by jumic** [comment]
@jameszh I think the correct name for the property is `ipSets` instead of `iPSets`.

```
networkfirewall.CfnRuleGroup(self, "RuleGroup", 
    type='STATEFUL',
    capacity = 20,
    rule_group_name="foo",
    rule_group={
        "rulesSource": {
        },
        "ruleVariables": {
            "ipSets": {
                "HOME_SET": {
                    "definition": [
                        "192.168.0.0/24"
                    ]
                }
            },
            "portSets": {
                "MYSQL": {
                    "definition": [
                        "3306"
                    ]
                }
            }
        }
    }
)
```

However, `cdk synth` causes this error:
```
jsii.errors.JavaScriptError: 
  Error: Value did not match any type in union: Value did not match any type in union: Value did not match any type in union: Value did not match any type in union: Expected object reference, got {"$jsii.map":{"definition":["192.168.0.0/24"]}}, Expected object reference, got {"$jsii.map":{"definition":["192.168.0.0/24"]}}, Expected object reference, got {"$jsii.map":{"HOME_SET":{"$jsii.map":{"definition":["192.168.0.0/24"]}}}}, Expected object reference, got {"$jsii.map":{"ipSets":{"$jsii.map":{"HOME_SET":{"$jsii.map":{"definition":["192.168.0.0/24"]}}}},"portSets":{"$jsii.map":{"MYSQL":{"$jsii.map":{"definition":["3306"]}}}}}}, Expected object reference, got {"$jsii.map":{"rulesSource":{"$jsii.map":{}},"ruleVariables":{"$jsii.map":{"ipSets":{"$jsii.map":{"HOME_SET":{"$jsii.map":{"definition":["192.168.0.0/24"]}}}},"portSets":{"$jsii.map":{"MYSQL":{"$jsii.map":{"definition":["3306"]}}}}}}}}
      at Kernel._toSandbox (/private/var/folders/zj/bc99wnlx4vj_h76zy2clrvh00000gn/T/tmp_ym4c482/lib/program.js:8513:23)
      at /private/var/folders/zj/bc99wnlx4vj_h76zy2clrvh00000gn/T/tmp_ym4c482/lib/program.js:9070:37
      at mapValues (/private/var/folders/zj/bc99wnlx4vj_h76zy2clrvh00000gn/T/tmp_ym4c482/lib/program.js:9341:35)
      at Object.deserialize (/private/var/folders/zj/bc99wnlx4vj_h76zy2clrvh00000gn/T/tmp_ym4c482/lib/program.js:9066:28)
      at Kernel._toSandbox (/private/var/folders/zj/bc99wnlx4vj_h76zy2clrvh00000gn/T/tmp_ym4c482/lib/program.js:8505:69)
      at /private/var/folders/zj/bc99wnlx4vj_h76zy2clrvh00000gn/T/tmp_ym4c482/lib/program.js:8553:42
      at Array.map (<anonymous>)
      at Kernel._boxUnboxParameters (/private/var/folders/zj/bc99wnlx4vj_h76zy2clrvh00000gn/T/tmp_ym4c482/lib/program.js:8553:27)
      at Kernel._toSandboxValues (/private/var/folders/zj/bc99wnlx4vj_h76zy2clrvh00000gn/T/tmp_ym4c482/lib/program.js:8539:29)
      at /private/var/folders/zj/bc99wnlx4vj_h76zy2clrvh00000gn/T/tmp_ym4c482/lib/program.js:8154:75

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/julian/dev/pythonCdk/app.py", line 15, in <module>
    PythonCdkStack(app, "PythonCdkStack",
  File "/Users/julian/dev/pythonCdk/.venv/lib/python3.9/site-packages/jsii/_runtime.py", line 83, in __call__
    inst = super().__call__(*args, **kwargs)
  File "/Users/julian/dev/pythonCdk/python_cdk/python_cdk_stack.py", line 63, in __init__
    networkfirewall.CfnRuleGroup(self, "RuleGroup", 
  File "/Users/julian/dev/pythonCdk/.venv/lib/python3.9/site-packages/jsii/_runtime.py", line 83, in __call__
    inst = super().__call__(*args, **kwargs)
  File "/Users/julian/dev/pythonCdk/.venv/lib/python3.9/site-packages/aws_cdk/aws_networkfirewall/__init__.py", line 1431, in __init__
    jsii.create(CfnRuleGroup, self, [scope, id, props])
  File "/Users/julian/dev/pythonCdk/.venv/lib/python3.9/site-packages/jsii/_kernel/__init__.py", line 275, in create
    response = self.provider.create(
  File "/Users/julian/dev/pythonCdk/.venv/lib/python3.9/site-packages/jsii/_kernel/providers/process.py", line 344, in create
    return self._process.send(request, CreateResponse)
  File "/Users/julian/dev/pythonCdk/.venv/lib/python3.9/site-packages/jsii/_kernel/providers/process.py", line 326, in send
    raise JSIIError(resp.error) from JavaScriptError(resp.stack)
jsii.errors.JSIIError: Value did not match any type in union: Value did not match any type in union: Value did not match any type in union: Value did not match any type in union: Expected object reference, got {"$jsii.map":{"definition":["192.168.0.0/24"]}}, Expected object reference, got {"$jsii.map":{"definition":["192.168.0.0/24"]}}, Expected object reference, got {"$jsii.map":{"HOME_SET":{"$jsii.map":{"definition":["192.168.0.0/24"]}}}}, Expected object reference, got {"$jsii.map":{"ipSets":{"$jsii.map":{"HOME_SET":{"$jsii.map":{"definition":["192.168.0.0/24"]}}}},"portSets":{"$jsii.map":{"MYSQL":{"$jsii.map":{"definition":["3306"]}}}}}}, Expected object reference, got {"$jsii.map":{"rulesSource":{"$jsii.map":{}},"ruleVariables":{"$jsii.map":{"ipSets":{"$jsii.map":{"HOME_SET":{"$jsii.map":{"definition":["192.168.0.0/24"]}}}},"portSets":{"$jsii.map":{"MYSQL":{"$jsii.map":{"definition":["3306"]}}}}}}}}
```

I did the same using the property classes:
```
networkfirewall.CfnRuleGroup(self, "RuleGroup", 
    type='STATEFUL',
    capacity = 20,
    rule_group_name="foo",
    rule_group=networkfirewall.CfnRuleGroup.RuleGroupProperty(
        rules_source=networkfirewall.CfnRuleGroup.RulesSourceProperty(),
        rule_variables=networkfirewall.CfnRuleGroup.RuleVariablesProperty(
            port_sets={
                "MYSQL": networkfirewall.CfnRuleGroup.PortSetProperty(
                    definition=[
                        "3306"
                    ]
                )
            },
            ip_sets={
                "HOME_SET": networkfirewall.CfnRuleGroup.IPSetProperty(
                    definition=[
                        "192.168.0.0/24"
                    ]
                ),
            }
        )
    )
)
```

Again, `cdk synth` fails:
```
Traceback (most recent call last):
  File "/Users/julian/dev/pythonCdk/app.py", line 15, in <module>
    PythonCdkStack(app, "PythonCdkStack",
  File "/Users/julian/dev/pythonCdk/.venv/lib/python3.9/site-packages/jsii/_runtime.py", line 83, in __call__
    inst = super().__call__(*args, **kwargs)
  File "/Users/julian/dev/pythonCdk/python_cdk/python_cdk_stack.py", line 35, in __init__
    "HOME_SET": networkfirewall.CfnRuleGroup.IPSetProperty(
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/typing.py", line 1072, in _no_init
    raise TypeError('Protocols cannot be instantiated')
TypeError: Protocols cannot be instantiated
Subprocess exited with error 1
```

This is caused by the `IPSetProperty`.
```
networkfirewall.CfnRuleGroup.IPSetProperty(
    definition=[
        "192.168.0.0/24"
    ]
)
```

```
Traceback (most recent call last):
  File "/Users/julian/dev/pythonCdk/app.py", line 15, in <module>
    PythonCdkStack(app, "PythonCdkStack",
  File "/Users/julian/dev/pythonCdk/.venv/lib/python3.9/site-packages/jsii/_runtime.py", line 83, in __call__
    inst = super().__call__(*args, **kwargs)
  File "/Users/julian/dev/pythonCdk/python_cdk/python_cdk_stack.py", line 15, in __init__
    networkfirewall.CfnRuleGroup.IPSetProperty(
  File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/typing.py", line 1072, in _no_init
    raise TypeError('Protocols cannot be instantiated')
TypeError: Protocols cannot be instantiated
Subprocess exited with error 1
```

Compared to `PortSetProperty`, using `PortSetProperty` works fine, no exception is thrown.
```
networkfirewall.CfnRuleGroup.PortSetProperty(
    definition=[
        "3306"
    ]
)
```

Probably the main issue is caused by differences between [IPSetProperty](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_networkfirewall/CfnRuleGroup.html#portsetproperty) and [PortSetProperty](https://docs.aws.amazon.com/cdk/api/latest/python/aws_cdk.aws_networkfirewall/CfnRuleGroup.html#ipsetproperty
).

From my point of view, it necessary to analyse why `networkfirewall.CfnRuleGroup.IPSetProperty` can't be initialized. Is this caused by `Bases: typing_extensions.Protocol` in the definition?

**Comment by jameszh** [comment]
I tried originally with ipSets, and it threw the same error (posted in Gitter). Now, I realize that it just silently passes any value  except correct one: ipSets (which is opposite of what I thought).

**Comment by skinny85** [comment]
Thanks for opening the issue @jameszh. Confirming I was able to reproduce it. This looks like a JSII problem, I've opened an issue to that project: https://github.com/aws/jsii/issues/2929, let's see what they say.

**Comment by jumic** [comment]
@skinny85 The problem should be fixed in #18275. You can probably close this issue.

**Comment by peterwoodworth** [comment]
You're right, the issue can be closed.
